### PR TITLE
Convert EXISTS/NOT EXISTS subquery into scalar subquery with LIMIT.

### PR DESCRIPTION
**Pull request by vraghavan78**
Before this PR, an EXISTS/NOT EXISTS subquery was converted into a
aggregate count(*) query, whose count it then checked to be greater
than 0 from EXISTS subquery, and = 0 for NOT EXISTS subquery.

The problem with this approach is that we need to compute the count
for the entire subquery even though that is a superfulous. All we need
to do is check if there is one tuple in the subquery or not.

Consider the input query to be `select * from customer where exists (select * from product);`

To ensure correctness, we need to make sure that there exists atleast one tuple in
subquery `(select * from product)`. To achieve this, we convert the above query as follows
`select * from customer where 0 < (select 1 as DUMMY_CONSTANT_ONE (select * from product limit 1));`
Here, the sole purpose of the DUMMY_CONSTANT_ONE is to validate that there is a single
tuple in the subquery

```
Input:
+--CLogicalSelect
   |--CLogicalGet "customer" ("customer"), Columns: ["cn" (13), "cname" (14), "cloc" (15), "ctid" (16), "xmin" (17), "cmin" (18), "xmax" (19), "cmax" (20), "tableoid" (21), "gp_segment_id" (22)] Key sets: {[0], [3,9]}
   +--CScalarSubqueryExists
       |--CLogicalGet "product" ("product"), Columns: ["pn" (23), "pname" (24), "pcolor" (25), "ctid" (26), "xmin" (27), "cmin" (28), "xmax" (29), "cmax" (30), "tableoid" (31), "gp_segment_id" (32)] Key sets: {[0], [3,9]}

Output:
+--CLogicalSelect
   |--CLogicalGet "customer" ("customer"), Columns: ["cn" (13), "cname" (14), "cloc" (15), "ctid" (16), "xmin" (17), "cmin" (18), "xmax" (19), "cmax" (20), "tableoid" (21), "gp_segment_id" (22)] Key sets: {[0], [3,9]}
   +--CScalarCmp (>)
      |--CScalarSubquery["ColRef_0035" (35)] generated by Exist SQ
      |  +--CLogicalProject
      |     |--CLogicalLimit <empty> global
      |     |  |--CLogicalGet "product" ("product"), Columns: ["pn" (23), "pname" (24), "pcolor" (25), "ctid" (26), "xmin" (27), "cmin" (28), "xmax" (29), "cmax" (30), "tableoid" (31), "gp_segment_id" (32)] Key sets: {[0], [3,9]}
      |     |  |--CScalarConst (0)
      |     |  +--CScalarConst (1)
      |     +--CScalarProjectList
      |        +--CScalarProjectElement "ColRef_0035" (35)
      |           +--CScalarConst (1)
      +--CScalarConst (0)
```

### CODE REVIEW

File: data/dxl/minidump/CorrelatedAntiSemiJoin-True.mdp [hunk 1]
```diff
@@ -865,7 +865,7 @@ SELECT pn, cn, vn FROM sale s WHERE NOT EXISTS (SELECT * FROM customer WHERE NOT
         </dxl:LogicalGet>
       </dxl:LogicalSelect>
     </dxl:Query>
-	<dxl:Plan Id="0" SpaceSize="4">
+	<dxl:Plan Id="0" SpaceSize="176">
```
Comment (hsyuan) [inline data/dxl/minidump/CorrelatedAntiSemiJoin-True.mdp]: Woa

**Comment by hsyuan** [comment]
For NOT EXISTS, I didn't see the changes:
```
hyuan=# explain select * from foo where exists (select * from t);
LOG:  statement: explain select * from foo where exists (select * from t);
                                                 QUERY PLAN
------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=0.00..1324032.19 rows=5 width=8)
   ->  Nested Loop EXISTS Join  (cost=0.00..1324032.19 rows=2 width=8)
         Join Filter: true
         ->  Table Scan on foo  (cost=0.00..431.00 rows=2 width=8)
         ->  Materialize  (cost=0.00..431.00 rows=1 width=1)
               ->  Broadcast Motion 1:3  (slice2)  (cost=0.00..431.00 rows=3 width=1)
                     ->  Limit  (cost=0.00..431.00 rows=1 width=1)
                           ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=1)
                                 ->  Table Scan on t  (cost=0.00..431.00 rows=1 width=1)
 Settings:  optimizer=on
 Optimizer status: PQO version 2.39.1
(11 rows)

hyuan=# explain select * from foo where not exists (select * from t);
LOG:  statement: explain select * from foo where not exists (select * from t);
                                            QUERY PLAN
---------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1324032.15 rows=5 width=8)
   ->  Nested Loop Left Anti Semi Join  (cost=0.00..1324032.15 rows=2 width=8)
         Join Filter: true
         ->  Table Scan on foo  (cost=0.00..431.00 rows=2 width=8)
         ->  Materialize  (cost=0.00..431.00 rows=1 width=1)
               ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=1)
                     ->  Table Scan on t  (cost=0.00..431.00 rows=1 width=1)
 Settings:  optimizer=on
 Optimizer status: PQO version 2.39.1
(9 rows)
```

**Comment by hsyuan** [comment]
I think it should go to expression preprocess step, because it is some kind of query rewrite.

**Comment by vraghavan78** [comment]
@hsyuan can u look at the discussion in https://github.com/greenplum-db/gporca/pull/193.

**Comment by vraghavan78** [comment]
This logic is busted when it comes to not-exists.

consider the following example:

```
create table foo(a int, b int);
create table bar(c int, d int);

insert into foo select i, i from generate_series(1,1000) i;
vraghavan=# drop table bar;
DROP TABLE
vraghavan=# create table bar(c int, d int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'c' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
vraghavan=# select min(a) from foo where not exists (select * from bar where b = 0);
 min
-----

(1 row)

vraghavan=# set optimizer =off;
SET
vraghavan=# select min(a) from foo where not exists (select * from bar where b = 0);
 min
-----
   1
(1 row)

```
Need to re-think this.

**Comment by vraghavan78** [comment]
The underlying issue is that for the case of NOT EXISTS we cannot just convert into `(COUNT = 0)` because COUNT of empty table is null. Therefore NOT EXISTS should be transformed into `((COUNT == 0) OR (COUNT IS NULL))`

Will open another PR with this logic.

**Pull request closed without merging**